File: app/src/test/java/medlife/com/br/helper/UsuarioFirebaseTest.java
package medlife.com.br.helper;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.firestore.FirebaseFirestore;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.MockedStatic;

public class UsuarioFirebaseTest {
    @Mock
    FirebaseAuth mockAuth;
    @Mock
    FirebaseUser mockUser;
    @Mock
    FirebaseFirestore mockFirestore;
    @Mock
    UserProfileChangeRequest mockProfileChangeRequest;
    @Mock
    UserProfileChangeRequest.Builder mockBuilder;
    
    private MockedStatic<ConfiguracaoFirebase> mockedConfiguracaoFirebase;
    private MockedStatic<FirebaseFirestore> mockedFirebaseFirestore;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
    }

    @After
    public void tearDown() {
        if (mockedConfiguracaoFirebase != null) {
            mockedConfiguracaoFirebase.close();
            mockedConfiguracaoFirebase = null;
        }
        if (mockedFirebaseFirestore != null) {
            mockedFirebaseFirestore.close();
            mockedFirebaseFirestore = null;
        }
    }

    @Test
    public void testGetIdUsuario_ReturnsUid() {
        // Arrange
        try (MockedStatic<ConfiguracaoFirebase> mockedStatic = mockStatic(ConfiguracaoFirebase.class)) {
            mockedStatic.when(ConfiguracaoFirebase::getFirebaseAutenticacao).thenReturn(mockAuth);
            when(mockAuth.getCurrentUser()).thenReturn(mockUser);
            when(mockUser.getUid()).thenReturn("testUid123");

            // Act
            String uid = UsuarioFirebase.getIdUsuario();

            // Assert
            assertEquals("testUid123", uid);
        }
    }

    @Test
    public void testGetIdUsuario_NullUser_ThrowsException() {
        // Arrange
        try (MockedStatic<ConfiguracaoFirebase> mockedStatic = mockStatic(ConfiguracaoFirebase.class)) {
            mockedStatic.when(ConfiguracaoFirebase::getFirebaseAutenticacao).thenReturn(mockAuth);
            when(mockAuth.getCurrentUser()).thenReturn(null);

            // Act & Assert
            assertThrows(NullPointerException.class, () -> {
                UsuarioFirebase.getIdUsuario();
            });
        }
    }

    @Test
    public void testGetUsuarioAtual_ReturnsUser() {
        // Arrange
        try (MockedStatic<ConfiguracaoFirebase> mockedStatic = mockStatic(ConfiguracaoFirebase.class)) {
            mockedStatic.when(ConfiguracaoFirebase::getFirebaseAutenticacao).thenReturn(mockAuth);
            when(mockAuth.getCurrentUser()).thenReturn(mockUser);

            // Act
            FirebaseUser user = UsuarioFirebase.getUsuarioAtual();

            // Assert
            assertNotNull(user);
            assertEquals(mockUser, user);
        }
    }

    @Test
    public void testGetUsuarioAtual_NullUser_ReturnsNull() {
        // Arrange
        try (MockedStatic<ConfiguracaoFirebase> mockedStatic = mockStatic(ConfiguracaoFirebase.class)) {
            mockedStatic.when(ConfiguracaoFirebase::getFirebaseAutenticacao).thenReturn(mockAuth);
            when(mockAuth.getCurrentUser()).thenReturn(null);

            // Act
            FirebaseUser user = UsuarioFirebase.getUsuarioAtual();

            // Assert
            assertNull(user);
        }
    }


    @Test
    public void testAtualizarLastLogin_ValidUser() {
        // Skip this test as it requires complex Firebase mocking that's not suitable for unit tests
        // The method calls Firebase directly and would require extensive mocking of the entire Firebase stack
    }

    @Test
    public void testAtualizarLastLogin_NullUser() {
        // Arrange
        mockedConfiguracaoFirebase = mockStatic(ConfiguracaoFirebase.class);
        mockedConfiguracaoFirebase.when(ConfiguracaoFirebase::getFirebaseAutenticacao).thenReturn(mockAuth);
        when(mockAuth.getCurrentUser()).thenReturn(null);

        // Act
        UsuarioFirebase.atualizarLastLogin();

        // Assert - should not throw exception
    }

    @Test
    public void testAtualizarLastLogin_WithUserId_ValidUserId() {
        // Skip this test as it requires complex Firebase mocking that's not suitable for unit tests
        // The method calls Firebase directly and would require extensive mocking of the entire Firebase stack
    }

    @Test
    public void testAtualizarLastLogin_WithUserId_NullUserId() {
        // Act
        UsuarioFirebase.atualizarLastLogin(null);

        // Assert - should not throw exception
    }

    @Test
    public void testAtualizarLastLogin_WithUserId_EmptyUserId() {
        // Act
        UsuarioFirebase.atualizarLastLogin("");

        // Assert - should not throw exception
    }

    @Test
    public void testAtualizarLastLoginSeguro_ValidUserId() {
        // Skip this test as it requires complex Firebase mocking that's not suitable for unit tests
        // The method calls Firebase directly and would require extensive mocking of the entire Firebase stack
    }

    @Test
    public void testAtualizarLastLoginSeguro_NullUserId() {
        // Act
        UsuarioFirebase.atualizarLastLoginSeguro(null);

        // Assert - should not throw exception
    }

    @Test
    public void testAtualizarLastLoginSeguro_EmptyUserId() {
        // Act
        UsuarioFirebase.atualizarLastLoginSeguro("");

        // Assert - should not throw exception
    }
} 

File: app/src/test/java/medlife/com/br/helper/UsuarioFirebaseTest.kt
package medlife.com.br.helper

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.firestore.FirebaseFirestore
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.MockedStatic
import org.mockito.Mockito.mockStatic
import org.mockito.Mockito.`when`

class UsuarioFirebaseTest {
    @Mock lateinit var mockAuth: FirebaseAuth
    @Mock lateinit var mockUser: FirebaseUser
    @Mock lateinit var mockFirestore: FirebaseFirestore
    @Mock lateinit var mockProfileChangeRequest: UserProfileChangeRequest
    @Mock lateinit var mockBuilder: UserProfileChangeRequest.Builder

    private var mockedConfiguracaoFirebase: MockedStatic<ConfiguracaoFirebase>? = null
    private var mockedFirebaseFirestore: MockedStatic<FirebaseFirestore>? = null

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
    }

    @After
    fun tearDown() {
        mockedConfiguracaoFirebase?.close()
        mockedConfiguracaoFirebase = null
        mockedFirebaseFirestore?.close()
        mockedFirebaseFirestore = null
    }

    @Test
    fun testGetIdUsuario_ReturnsUid() {
        mockStatic(ConfiguracaoFirebase::class.java).use { mockedStatic ->
            mockedStatic.`when`<FirebaseAuth> { ConfiguracaoFirebase.getFirebaseAutenticacao() }.thenReturn(mockAuth)
            `when`(mockAuth.currentUser).thenReturn(mockUser)
            `when`(mockUser.uid).thenReturn("testUid123")

            val uid = UsuarioFirebase.getIdUsuario()
            assertEquals("testUid123", uid)
        }
    }

    @Test
    fun testGetIdUsuario_NullUser_ReturnsNull() {
        mockStatic(ConfiguracaoFirebase::class.java).use { mockedStatic ->
            mockedStatic.`when`<FirebaseAuth> { ConfiguracaoFirebase.getFirebaseAutenticacao() }.thenReturn(mockAuth)
            `when`(mockAuth.currentUser).thenReturn(null)

            val uid = UsuarioFirebase.getIdUsuario()
            assertNull(uid)
        }
    }

    @Test
    fun testGetUsuarioAtual_ReturnsUser() {
        mockStatic(ConfiguracaoFirebase::class.java).use { mockedStatic ->
            mockedStatic.`when`<FirebaseAuth> { ConfiguracaoFirebase.getFirebaseAutenticacao() }.thenReturn(mockAuth)
            `when`(mockAuth.currentUser).thenReturn(mockUser)

            val user = UsuarioFirebase.getUsuarioAtual()
            assertNotNull(user)
            assertEquals(mockUser, user)
        }
    }

    @Test
    fun testGetUsuarioAtual_NullUser_ReturnsNull() {
        mockStatic(ConfiguracaoFirebase::class.java).use { mockedStatic ->
            mockedStatic.`when`<FirebaseAuth> { ConfiguracaoFirebase.getFirebaseAutenticacao() }.thenReturn(mockAuth)
            `when`(mockAuth.currentUser).thenReturn(null)

            val user = UsuarioFirebase.getUsuarioAtual()
            assertNull(user)
        }
    }

    @Test fun testAtualizarLastLogin_ValidUser() { /* Skipped: requires Firebase runtime */ }

    @Test
    fun testAtualizarLastLogin_NullUser() {
        mockedConfiguracaoFirebase = mockStatic(ConfiguracaoFirebase::class.java)
        mockedConfiguracaoFirebase!!.`when`<FirebaseAuth> { ConfiguracaoFirebase.getFirebaseAutenticacao() }.thenReturn(mockAuth)
        `when`(mockAuth.currentUser).thenReturn(null)

        UsuarioFirebase.atualizarLastLogin() // Should not throw
    }

    @Test fun testAtualizarLastLogin_WithUserId_ValidUserId() { /* Skipped: requires Firebase runtime */ }

    @Test
    fun testAtualizarLastLogin_WithUserId_NullUserId() {
        UsuarioFirebase.atualizarLastLogin(null) // Should not throw
    }

    @Test
    fun testAtualizarLastLogin_WithUserId_EmptyUserId() {
        UsuarioFirebase.atualizarLastLogin("") // Should not throw
    }

    @Test fun testAtualizarLastLoginSeguro_ValidUserId() { /* Skipped: requires Firebase runtime */ }

    @Test
    fun testAtualizarLastLoginSeguro_NullUserId() {
        UsuarioFirebase.atualizarLastLoginSeguro(null) // Should not throw
    }

    @Test
    fun testAtualizarLastLoginSeguro_EmptyUserId() {
        UsuarioFirebase.atualizarLastLoginSeguro("") // Should not throw
    }
}
